backend/de.metas.banking/de.metas.banking.base/src/main/java/de/metas/banking/service/RemittanceAdviceBankingService.java

```diff
@@ -0,0 +1,350 @@
+					remittanceAdvice.getRemittanceAdviceId().getRepoId(), remittanceAdvice.getDocStatus(), remittanceAdvice.getPaymentId());
+		}
+	}
+
+	@NonNull
+	private I_C_Payment createPayment(@NonNull final RemittanceAdvice remittanceAdvice)
+	{
+		final DefaultPaymentBuilder paymentBuilder = remittanceAdvice.isSOTrx()
+				? paymentBL.newInboundReceiptBuilder()
+				: paymentBL.newOutboundPaymentBuilder();
+
+		final BPartnerBankAccountId bPartnerBankAccountId = remittanceAdvice.isSOTrx() ? remittanceAdvice.getSourceBPartnerBankAccountId()
+				: remittanceAdvice.getDestinationBPartnerBankAccountId();
+
+		final ZoneId tz = orgDAO.getTimeZone(remittanceAdvice.getOrgId());
+
+		return paymentBuilder
+				.adOrgId(remittanceAdvice.getOrgId())
+				.bpartnerId(remittanceAdvice.isSOTrx() ? remittanceAdvice.getSourceBPartnerId() : remittanceAdvice.getDestinationBPartnerId())
+				.orgBankAccountId(BankAccountId.ofRepoId(bPartnerBankAccountId.getRepoId()))
+				.currencyId(remittanceAdvice.getRemittedAmountCurrencyId())
+				.payAmt(remittanceAdvice.getRemittedAmountSum())
+				.dateAcct(CoalesceUtil.coalesce(TimeUtil.asLocalDate(remittanceAdvice.getPaymentDate(), tz),
+						TimeUtil.asLocalDate(remittanceAdvice.getDocumentDate(), tz)))
+				.dateTrx(TimeUtil.asLocalDate(remittanceAdvice.getDocumentDate(), tz))
+				.tenderType(TenderType.DirectDeposit)
+				.createAndProcess();
+	}
+
+	@NonNull
+	private PaymentAllocationCriteria getPaymentAllocationCriteria(
+			@NonNull final RemittanceAdvice remittanceAdvice,
+			@NonNull final I_C_Payment payment)
+	{
+		final Map<InvoiceId, I_C_Invoice> invoiceMapById = getInvoiceMapById(remittanceAdvice.getLines());
+
+		final List<PaymentAllocationPayableItem> paymentAllocationPayableItems = remittanceAdvice.getLines()
+				.stream()
+				.map(line -> createPaymentAllocationPayableItem(line, remittanceAdvice, invoiceMapById.get(line.getInvoiceId())))
+				.collect(Collectors.toList());
+
+		return PaymentAllocationCriteria.builder()
+				.payment(payment)
+				.dateTrx(remittanceAdvice.getSendDate() != null ? remittanceAdvice.getSendDate() : SystemTime.asInstant())
+				.paymentAllocationPayableItems(paymentAllocationPayableItems)
+				.allowPartialAllocations(true)
+				.allowInvoiceToCreditMemoAllocation(false) // invoices and creditmemos of the remadv must not be allocated against each other!
+				.build();
+	}
+
+	private void populateRemittanceWithAllocationData(
+			@NonNull final Map<InvoiceId, RemittanceAdviceLine> remittanceAdviceLineMap,
+			@NonNull final PaymentAllocationResult paymentAllocationResult)
+	{
+		final ImmutableSet<AllocationLineCandidate.AllocationLineCandidateType> validTypes =
+				ImmutableSet.of(AllocationLineCandidate.AllocationLineCandidateType.SalesInvoiceToPurchaseInvoice,
+						AllocationLineCandidate.AllocationLineCandidateType.SalesCreditMemoToPurchaseInvoice);
+
+		final Map<Integer, InvoiceId> serviceFeeInvoiceIdsByAssignedInvoiceId =
```

Comment:
i find this method very hard to read. consider breaking it up in several methods, e.g.
* the code that creates serviceFeeInvoiceIdsByAssignedInvoiceId can be one method. Instead of Integer index, it could be TableRecordReference, PaymentId or something else but descriptive.
* the .forEach(...) can be another method